### PR TITLE
Bug 1872964 - Add FCM testing documentation

### PR DESCRIPTION
**Pull request by jonalmeida**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1872964

### CODE REVIEW

**Review by Amejia481 — APPROVED** [review]
📝 ✅

**Review by mhammond — APPROVED** [review]
Thanks Jon! These instructions appear to work fine. I found some of the FAQ a little confusing, so please feel free to take or ignore those comments as you see fit!

File: fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+## Firefox Sync & Send Tab
+If you are testing WebPush support with Firefox Sync, you should also expect to see this log shortly after signing into a Firefox Sync account:
+```
+Created a new subscription: [context data]
+```
+
+## FAQ
+
+**Q. Why are credentials stored in Google Drive?**
+  * According to [Firebase engineers](https://groups.google.com/forum/#!msg/firebase-talk/bamCgTDajkw/uVEJXjtiBwAJ), it should be safe to commit our `google-services.json` file which contains our configuration of Firebase push service. In fact, the Google I/O app commits [their release version](https://github.com/google/iosched/blob/b428d2be4bb96bd423e47cb709c906ce5d02150f/mobile/google-services.json) of the file to Github as well. For our case, we can use the generated XML file created from the `google-services.json`.
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md:31]: Ah, I tried to answer this in the question below, but maybe I should combine the two and expand.

[the remark this answers is not in the document]
Comment (jonalmeida) [reply fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md:31]: Fair point! I didn't really explain what the `google-services.json` is and where it comes from.

Your bullets are correct, but it doesn't sound good that it wasn't as clear from a first read, so I'll re-word this.

File: fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+2. Place the file in the directory `fenix/app/src/debug/res/values/`.
+    * This will be compiled into the `fenixDebug` variant.
+3. Run the app and verify you receive log messages in `adb logcat` like the ones below:
+    ```
+    Received a new registration token from push service.
+    Got new Firebase token.
+    ```
+
+## Components and code
+The important components for push to work are the `feature-push` (specifically look at the [README][3]) and `lib-push-firebase`.
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md:16]: Sure!

**Comment by mhammond** [comment]
Possibly a topic for another PR, but after adding this file, android studio offered to add the file to git, including an "always add" option. `git status` is also obviously showing this file. I'm wondering if this file should be added to `.gitignore`?

**Comment by jonalmeida** [comment]
> Possibly a topic for another PR, but after adding this file, android studio offered to add the file to git, including an "always add" option. `git status` is also obviously showing this file. I'm wondering if this file should be added to `.gitignore`?

Right, we could try adding a gitignore if we know the filename will not change and location is fine. You can technically name the file anything because all the XML resource files are compiled into one in the end. We also do not have to place it in the `debug/values` path as it's only my recommendation for having the file included in only a debug variant.

I'm unsure what a reasonable gitignore expression would look like then. Open to ideas!

**Comment by mhammond** [comment]
.gitignore isn't a huge issue, but I think it would be fine to assume people have followed these instructions and know what they are doing if they did something different.  IOW, I'd think that `fenix/app/src/**/res/values/fenix_firebase_push_credentials.xml` would be sufficient.

**Comment by jonalmeida** [comment]
Oh yes, I forget about `**`. Let me add that into this PR in that case.

File: fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+## Firefox Sync & Send Tab
+If you are testing WebPush support with Firefox Sync, you should also expect to see this log shortly after signing into a Firefox Sync account:
+```
+Created a new subscription: [context data]
+```
+
+## FAQ
+
+**Q. Why are credentials stored in Google Drive?**
+  * Firebase services require a credential file, typically named `google-services.json`, which are retrieved from the Firebase console for all the applications in that project. Firebase uses the `com.google.gms.google-services` Gradle plugin in order to generate an XML file to consume with their SDK for the Android application that matches the app ID in that file (this can have any name).
```
Comment (mhammond) [inline fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md:31]: TBH, I think this is still a bit wordy - how about something like:

Firebase services require a credential file, typically named google-services.json, which is retrieved from the Firebase console for all the applications in that project. Our requirements are somewhat unique, so we generate our own XML file from google-services.json and use that. According to [Firebase engineers](https://groups.google.com/forum/#!msg/firebase-talk/bamCgTDajkw/uVEJXjtiBwAJ), it should be safe to commit the info in this file, but we avoid doing that as there are other forks of Fenix that must not use these credentials.

File: fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    According to [Firebase engineers](https://groups.google.com/forum/#!msg/firebase-talk/bamCgTDajkw/uVEJXjtiBwAJ), it should be safe to commit our `google-services.json` file which contains our configuration of Firebase Cloud Messaging. In fact, the Google I/O app commits [their release version](https://github.com/google/iosched/blob/b428d2be4bb96bd423e47cb709c906ce5d02150f/mobile/google-services.json) of the file to Github as well. 
+    While it is safe to do so, we would not advise it as there are other forks of Fenix that should not mistakenly use these credentials in error. For our case, we use the generated XML file directly in our applications that were created from the `google-services.json`.
+
+**Q. Why can we not add the JSON file (mentioned above) directly in the app?**
```
Comment (mhammond) [inline fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md:35]: and this could be Q. What are the special requirement which mean we can't directly use google-services.json?

**Review by mhammond — APPROVED** [review]

File: .gitignore [hunk 5]
```diff
@@ -72,6 +72,7 @@ fastlane/
 .nimbus
 .wallpaper_url
 .pocket_consumer_key
+fenix/app/src/**/res/values/fenix_firebase_push_credentials.xml
```
Comment (mhammond) [inline .gitignore:75]: FWIW, I just verified this works for me, thanks :)

File: fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md (hunk 3 above)
Comment (jonalmeida) [reply fenix/docs/Firebase-Cloud-Messaging-for-WebPush.md]: Thanks, this looks good!

I had added the "file which contains our configuration of Firebase Cloud Messaging." because you get a JSON file of all applications (Fenix, Focus, Reference Browser) and I wanted to make sure it's clear that we don't want to mistakenly use the wrong config in that file. That said, maybe it's extra details.

I did leave the Google I/O app example in there as proof of a Google app that follows the advice dished out.

Learn lesson: don't write docs at the end of the day or when hungry. 😄
